polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -229,50 +240,125 @@ pub fn worker_entrypoint(
 					},
 				};
 
-				// SAFETY: new process is spawned within a single threaded process. This invariant
-				// is enforced by tests.
-				let result = match unsafe { nix::unistd::fork() } {
-					Err(errno) => Err(error_from_errno("fork", errno)),
-					Ok(ForkResult::Child) => {
-						// Dropping the stream closes the underlying socket. We want to make sure
-						// that the sandboxed child can't get any kind of information from the
-						// outside world. The only IPC it should be able to do is sending its
-						// response over the pipe.
-						drop(stream);
-						// Drop the read end so we don't have too many FDs open.
-						drop(pipe_reader);
-
-						handle_child_process(
-							pvf,
-							pipe_writer,
-							preparation_timeout,
-							prepare_job_kind,
-							executor_params,
-						)
-					},
-					Ok(ForkResult::Parent { child }) => {
-						// the read end will wait until all write ends have been closed,
-						// this drop is necessary to avoid deadlock
-						drop(pipe_writer);
-
-						handle_parent_process(
-							pipe_reader,
-							worker_pid,
-							child,
-							temp_artifact_dest.clone(),
-							usage_before,
-							preparation_timeout,
-						)
-					},
-				};
-
-				gum::trace!(
-					target: LOG_TARGET,
-					%worker_pid,
-					"worker: sending result to host: {:?}",
-					result
-				);
-				send_response(&mut stream, result)?;
+				let stream_fd = stream.as_raw_fd();
+
+				cfg_if::cfg_if! {
+					if #[cfg(target_os = "linux")] {
+						let mut stack: Vec<u8> = vec![0u8; EXECUTE_THREAD_STACK_SIZE];
+						let flags = CloneFlags::CLONE_NEWCGROUP
+						| CloneFlags::CLONE_NEWIPC
+						| CloneFlags::CLONE_NEWNET
+						| CloneFlags::CLONE_NEWNS
+						| CloneFlags::CLONE_NEWPID
+						| CloneFlags::CLONE_NEWUSER
+						| CloneFlags::CLONE_NEWUTS
+
+						// SIGCHLD flag is used to inform clone that the parent process is
+						// expecting a child termination signal, without this flag `waitpid` function
+						 // return `ECHILD` error.
+						| CloneFlags::from_bits_retain(libc::SIGCHLD);
+
+						// SAFETY: new process is spawned within a single threaded process. This invariant
+						// is enforced by tests. Stack size being specified to ensure child doesn't overflow
+						let result = match unsafe { nix::sched::clone(
+							Box::new(|| {
+								// SAFETY: pipe_writer is an open and owned file descriptor at this point.
+								let mut pipe_writer = PipeWriter::from_raw_fd(pipe_writer);
+
+								if let Err(errno) = nix::unistd::close(pipe_reader) {
+									send_child_response(&mut pipe_writer, JobResult::Err(error_from_errno("closing pipe", errno)));
+								}
+
+								if let Err(errno) = nix::unistd::close(stream_fd) {
+									send_child_response(&mut pipe_writer, JobResult::Err(error_from_errno("error closing stream", errno)));
```

Comment:
Can we move the dropping code to the `handle_*_process` functions to clean up some duplication here? We can just pass in raw fd's to those functions, with comments to ensure on the caller-side that the correct fd's are given. This way we won't need the `PipeWriter`/`PipeReader` abstractions from `os_pipe` and can remove that dependency.

Reply:
Random note: one might wonder (or at least, I was wondering), instead of using the raw fd's, why not call `try_clone()` on `PipeWriter`/`PipeReader` before fork/clone? But that invokes [`dup`](https://man7.org/linux/man-pages/man2/dup.2.html) under the hood, which "allocates a new file descriptor that refers to the same open file description as the descriptor oldfd". Closing the cloned fd from `dup` in the child would not accomplish much, because the original fd would still be present in the child's address space (though, in the case of `clone`, not accessible by normal Rust code due to the borrow checker).